File: netraven/web/frontend/src/store/tags.js
import { defineStore } from 'pinia'
import { tagService } from '../api/api'

// Create a single default tag
const DEFAULT_TAGS = [
  {
    id: 'tag-default',
    name: 'Default',
    color: '#6366F1', // Indigo color
    description: 'Default tag for all devices',
    created_at: new Date().toISOString(),
    updated_at: new Date().toISOString()
  }
];

export const useTagStore = defineStore('tags', {
  state: () => ({
    tags: [...DEFAULT_TAGS], // Initialize with just the default tag
    currentTag: null,
    deviceTags: {},  // Map of device ID -> tags for that device
    loading: false,
    error: null
  }),
  
  getters: {
    getTagById: (state) => (id) => {
      return state.tags.find(tag => tag.id === id)
    },
    
    getTagByName: (state) => (name) => {
      return state.tags.find(tag => tag.name === name)
    },
    
    getTagsForDevice: (state) => (deviceId) => {
      return state.deviceTags[deviceId] || []
    }
  },
  
  actions: {
    async fetchTags() {
      this.loading = true
      this.error = null
      
      try {
        // Call the API to get all tags
        console.log('Fetching all tags')
        const tags = await tagService.getTags()
        
        // Update the store with the tags from the API
        this.tags = tags
        
        // Ensure we have at least the default tag if the API returns no tags
        if (this.tags.length === 0) {
          this.tags = [...DEFAULT_TAGS]
        }
        
        return this.tags
      } catch (error) {
        this.error = error.message || 'Failed to fetch tags'
        console.error('Error fetching tags:', error)
        return []
      } finally {
        this.loading = false
      }
    },
    
    async fetchTag(id) {
      this.loading = true
      this.error = null
      
      try {
        console.log(`Fetching tag with ID: ${id}`)
        const tag = await tagService.getTag(id)
        this.currentTag = tag
        
        // Update the tag in the tags array if it exists
        const index = this.tags.findIndex(t => t.id === id)
        if (index !== -1) {
          this.tags[index] = tag
        } else {
          this.tags.push(tag)
        }
        
        return tag
      } catch (error) {
        // Check if it's a 404 not found error
        if (error.response && error.response.status === 404) {
          console.error(`Tag with ID ${id} was not found in the database`)
          this.error = `Tag with ID ${id} not found`
        } else {
          this.error = error.response?.data?.detail || `Failed to fetch tag ${id}`
          console.error(`Error fetching tag ${id}:`, error)
        }
        return null
      } finally {
        this.loading = false
      }
    },
    
    async fetchDevicesForTag(id) {
      this.loading = true
      this.error = null
      
      try {
        const devices = await tagService.getDevicesForTag(id)
        return devices
      } catch (error) {
        this.error = error.response?.data?.detail || `Failed to fetch devices for tag ${id}`
        console.error(`Error fetching devices for tag ${id}:`, error)
        return []
      } finally {
        this.loading = false
      }
    },
    
    async fetchTagsForDevice(deviceId) {
      this.loading = true
      this.error = null
      
      console.log(`=== STORE DEBUG: fetchTagsForDevice ===`)
      console.log(`Method called with deviceId:`, deviceId)
      console.log(`deviceId type:`, typeof deviceId)
      
      try {
        console.log(`Fetching tags for device ${deviceId}`)
        
        // Instead of checking each tag for the device,
        // use the device/{deviceId}/tags endpoint directly
        const deviceTags = await tagService.getTagsForDevice(deviceId)
        console.log(`Got tags for device ${deviceId}:`, deviceTags)
        
        // Update our local cache
        this.deviceTags[deviceId] = deviceTags
        console.log(`Updated deviceTags in store:`, this.deviceTags)
        console.log(`=== END STORE DEBUG ===`)
        
        return deviceTags
      } catch (error) {
        this.error = error.response?.data?.detail || `Failed to fetch tags for device ${deviceId}`
        console.error(`Error fetching tags for device ${deviceId}:`, error)
        console.error('Error details:', {
          message: error.message,
          response: error.response?.data,
          status: error.response?.status
        })
        
        // If API call fails, use cached data if available
        if (this.deviceTags[deviceId] && this.deviceTags[deviceId].length > 0) {
          console.log(`Using cached tags for device ${deviceId}`)
          return this.deviceTags[deviceId]
        }
        
        return []
      } finally {
        this.loading = false
      }
    },
    
    async createTag(tagData) {
      this.loading = true
      this.error = null
      
      try {
        console.log('Creating new tag with data:', tagData)
        
        // Call the API to create the tag
        const newTag = await tagService.createTag(tagData)
        
        // Add the server-created tag to our local state
        this.tags.push(newTag)
        
        console.log('Created new tag:', newTag)
        return newTag
      } catch (error) {
        this.error = error.response?.data?.detail || error.message || 'Failed to create tag'
        console.error('Error creating tag:', error)
        return null
      } finally {
        this.loading = false
      }
    },
    
    async updateTag(id, tagData) {
      this.loading = true
      this.error = null
      
      try {
        console.log(`Updating tag ${id} with data:`, tagData)
        
        // Call the API to update the tag
        const updatedTag = await tagService.updateTag(id, tagData)
        
        // Update the tag in the tags array
        const index = this.tags.findIndex(t => t.id === id)
        if (index !== -1) {
          this.tags[index] = updatedTag
        }
        
        // Update currentTag if it matches
        if (this.currentTag && this.currentTag.id === id) {
          this.currentTag = updatedTag
        }
        
        return updatedTag
      } catch (error) {
        // Check if it's a 404 not found error
        if (error.response && error.response.status === 404) {
          console.error(`Tag with ID ${id} was not found in the database`)
          this.error = `Tag with ID ${id} not found`
        } else {
          this.error = error.response?.data?.detail || `Failed to update tag ${id}`
          console.error(`Error updating tag ${id}:`, error)
        }
        return null
      } finally {
        this.loading = false
      }
    },
    
    async deleteTag(id) {
      this.loading = true
      this.error = null
      
      try {
        // Call the API to delete the tag
        await tagService.deleteTag(id)
        
        // Remove the tag from the tags array
        this.tags = this.tags.filter(t => t.id !== id)
        
        // Clear currentTag if it matches
        if (this.currentTag && this.currentTag.id === id) {
          this.currentTag = null
        }
        
        // Remove from deviceTags
        for (const deviceId in this.deviceTags) {
          this.deviceTags[deviceId] = this.deviceTags[deviceId].filter(t => t.id !== id)
        }
        
        console.log(`Successfully deleted tag ${id}`)
        return true
      } catch (error) {
        // Check if it's a 404 not found error
        if (error.response && error.response.status === 404) {
          console.error(`Tag with ID ${id} was not found in the database`)
          
          // Still remove from local state if it doesn't exist on the server
          this.tags = this.tags.filter(t => t.id !== id)
          
          if (this.currentTag && this.currentTag.id === id) {
            this.currentTag = null
          }
          
          // Return success since we've removed it locally
          return true
        }
        
        this.error = error.response?.data?.detail || `Failed to delete tag ${id}`
        console.error(`Error deleting tag ${id}:`, error)
        return false
      } finally {
        this.loading = false
      }
    },
    
    async assignTagsToDevices(deviceIds, tagIds) {
      this.loading = true
      this.error = null
      
      try {
        const result = await tagService.assignTagsToDevices(deviceIds, tagIds)
        
        // Update deviceTags cache for each device
        for (const deviceId of deviceIds) {
          // Remove existing cache for this device to force refresh
          delete this.deviceTags[deviceId]
        }
        
        return result
      } catch (error) {
        this.error = error.response?.data?.detail || 'Failed to assign tags to devices'
        console.error('Error assigning tags to devices:', error)
        return null
      } finally {
        this.loading = false
      }
    },
    
    async removeTagsFromDevices(deviceIds, tagIds) {
      this.loading = true
      this.error = null
      
      try {
        const result = await tagService.removeTagsFromDevices(deviceIds, tagIds)
        
        // Update deviceTags cache for each device
        for (const deviceId of deviceIds) {
          // Remove existing cache for this device to force refresh
          delete this.deviceTags[deviceId]
        }
        
        return result
      } catch (error) {
        this.error = error.response?.data?.detail || 'Failed to remove tags from devices'
        console.error('Error removing tags from devices:', error)
        return null
      } finally {
        this.loading = false
      }
    },
    
    // Add single tag to a single device
    async assignTagToDevice(deviceId, tagId) {
      this.loading = true
      this.error = null
      
      try {
        console.log(`Assigning tag ${tagId} to device ${deviceId}`)
        
        // Add more detailed logging
        console.log('API call parameters:', {
          deviceIds: [deviceId],
          tagIds: [tagId]
        })
        
        // Call the API to assign the tag to the device
        const result = await tagService.assignTagsToDevices([deviceId], [tagId])
        
        console.log('API response:', result)
        
        if (!result) {
          throw new Error(`Failed to assign tag ${tagId} to device ${deviceId}: No response from API`)
        }
        
        if (!result.success) {
          console.error('Assignment failed with result:', result)
          throw new Error(result.message || `Failed to assign tag ${tagId} to device ${deviceId}`)
        }
        
        console.log('Tag assignment successful, refreshing device tags')
        
        // Refresh the device tags from the server to ensure we have the latest data
        await this.fetchTagsForDevice(deviceId)
        
        // Return the updated tags
        return this.deviceTags[deviceId] || []
      } catch (error) {
        const errorMessage = error.response?.data?.detail || error.message || `Failed to assign tag ${tagId} to device ${deviceId}`
        this.error = errorMessage
        console.error(`Error assigning tag ${tagId} to device ${deviceId}:`, error)
        console.error('Error details:', {
          message: error.message,
          response: error.response?.data,
          status: error.response?.status
        })
        throw error
      } finally {
        this.loading = false
      }
    },
    
    // Remove a tag from a device
    async removeTagFromDevice(deviceId, tagId) {
      this.loading = true
      this.error = null
      
      try {
        console.log(`Removing tag ${tagId} from device ${deviceId}`)
        
        // Add more detailed logging
        console.log('API call parameters:', {
          deviceIds: [deviceId],
          tagIds: [tagId]
        })
        
        // Call the API to remove the tag from the device
        const result = await tagService.removeTagsFromDevices([deviceId], [tagId])
        
        console.log('API response:', result)
        
        if (!result) {
          throw new Error(`Failed to remove tag ${tagId} from device ${deviceId}: No response from API`)
        }
        
        if (!result.success) {
          console.error('Removal failed with result:', result)
          throw new Error(result.message || `Failed to remove tag ${tagId} from device ${deviceId}`)
        }
        
        console.log('Tag removal successful, refreshing device tags')
        
        // Refresh the device tags from the server to ensure we have the latest data
        await this.fetchTagsForDevice(deviceId)
        
        // Return the updated tags
        return this.deviceTags[deviceId] || []
      } catch (error) {
        const errorMessage = error.response?.data?.detail || error.message || `Failed to remove tag ${tagId} from device ${deviceId}`
        this.error = errorMessage
        console.error(`Error removing tag ${tagId} from device ${deviceId}:`, error)
        console.error('Error details:', {
          message: error.message,
          response: error.response?.data,
          status: error.response?.status
        })
        throw error
      } finally {
        this.loading = false
      }
    }
  }
}) 

File: netraven/web/routers/tags.py
"""
Tags router for managing device tags.

This module provides endpoints for creating, retrieving, updating, and 
deleting tags for network devices. It also supports operations for 
retrieving devices associated with tags.
"""

from fastapi import APIRouter, Depends, HTTPException, Query, Path, status
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List, Optional, Dict, Any

from netraven.web.database import get_async_session
from netraven.web.auth import UserPrincipal, get_current_principal
from netraven.web.auth.permissions import require_scope, require_admin

# Schemas
from netraven.web.schemas.tag import (
    Tag as TagSchema, 
    TagCreate, 
    TagUpdate, 
    TagAssignment, 
    TagRemoval
)
from netraven.web.schemas.device import Device as DeviceSchema

# Logging
from netraven.core.logging import get_logger

# Factory and Services
from netraven.core.services.service_factory import ServiceFactory

logger = get_logger(__name__)
router = APIRouter(prefix="/tags", tags=["tags"])

# --- Tag CRUD --- 

@router.get("/", response_model=List[TagSchema])
async def list_tags(
    skip: int = Query(0, description="Number of records to skip", ge=0),
    limit: int = Query(100, description="Maximum number of records to return", ge=1, le=1000),
    principal: UserPrincipal = Depends(require_scope("read:tags")),
    session: AsyncSession = Depends(get_async_session),
    factory: ServiceFactory = Depends(ServiceFactory)
) -> List[TagSchema]:
    """
    List all tags.
    
    This endpoint requires the read:tags scope.
    """
    try:
        tags = await factory.tag_service.list_tags(skip=skip, limit=limit)
        
        logger.info(f"Access granted: user={principal.username}, resource=tags, action=list, count={len(tags)}")
        return tags
    except Exception as e:
        logger.error(f"Error listing tags: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error listing tags: {str(e)}"
        )

@router.post("/", response_model=TagSchema, status_code=status.HTTP_201_CREATED)
async def create_tag(
    tag_data: TagCreate,
    principal: UserPrincipal = Depends(require_scope("write:tags")),
    session: AsyncSession = Depends(get_async_session),
    factory: ServiceFactory = Depends(ServiceFactory)
) -> TagSchema:
    """
    Create a new tag.
    
    This endpoint requires the write:tags scope.
    """
    try:
        new_tag = await factory.tag_service.create_tag(tag_data)
        
        logger.info(f"Tag created: id={new_tag.id}, name={new_tag.name}, user={principal.username}")
        return new_tag
    except Exception as e:
        logger.error(f"Error creating tag: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Error creating tag: {str(e)}"
        )

@router.get("/{tag_id}", response_model=TagSchema)
async def get_tag(
    tag_id: str,
    principal: UserPrincipal = Depends(require_scope("read:tags")),
    session: AsyncSession = Depends(get_async_session),
    factory: ServiceFactory = Depends(ServiceFactory)
) -> TagSchema:
    """
    Get a specific tag by ID.
    
    This endpoint requires the read:tags scope.
    """
    try:
        tag = await factory.tag_service.get_tag_by_id(tag_id)
        
        if not tag:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Tag with ID {tag_id} not found"
            )
        
        logger.info(f"Access granted: user={principal.username}, resource=tag:{tag_id}, action=get")
        return tag
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error retrieving tag {tag_id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error retrieving tag: {str(e)}"
        )

@router.put("/{tag_id}", response_model=TagSchema)
async def update_tag(
    tag_id: str,
    tag_data: TagUpdate,
    principal: UserPrincipal = Depends(require_scope("write:tags")),
    session: AsyncSession = Depends(get_async_session),
    factory: ServiceFactory = Depends(ServiceFactory)
) -> TagSchema:
    """
    Update a tag.
    
    This endpoint requires the write:tags scope.
    """
    try:
        updated_tag = await factory.tag_service.update_tag(tag_id, tag_data)
        
        if not updated_tag:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Tag with ID {tag_id} not found"
            )
            
        logger.info(f"Tag updated: id={tag_id}, user={principal.username}")
        return updated_tag
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating tag {tag_id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Error updating tag: {str(e)}"
        )

@router.delete("/{tag_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_tag(
    tag_id: str,
    principal: UserPrincipal = Depends(require_scope("delete:tags")),
    session: AsyncSession = Depends(get_async_session),
    factory: ServiceFactory = Depends(ServiceFactory)
):
    """
    Delete a tag.
    
    This endpoint requires the delete:tags scope.
    """
    try:
        result = await factory.tag_service.delete_tag(tag_id)
        
        if not result:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Tag with ID {tag_id} not found"
            )
        
        logger.info(f"Tag deleted: id={tag_id}, user={principal.username}")
        return None
    except HTTPException as http_exc:
        if http_exc.status_code == status.HTTP_400_BAD_REQUEST:
            logger.warning(f"Tag deletion failed: id={tag_id}, user={principal.username}, reason={http_exc.detail}")
        raise http_exc
    except Exception as e:
        logger.error(f"Error deleting tag {tag_id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error deleting tag: {str(e)}"
        )

# --- Tag Association Endpoints --- 

@router.get("/{tag_id}/devices", response_model=List[DeviceSchema])
async def get_devices_for_tag(
    tag_id: str = Path(..., description="The ID of the tag"),
    skip: int = Query(0, description="Number of records to skip", ge=0),
    limit: int = Query(100, description="Maximum number of records to return", ge=1, le=1000),
    principal: UserPrincipal = Depends(require_scope(["read:tags", "read:devices"])),
    session: AsyncSession = Depends(get_async_session),
    factory: ServiceFactory = Depends(ServiceFactory)
) -> List[DeviceSchema]:
    """
    Get devices associated with a specific tag.
    
    This endpoint requires both read:tags and read:devices scopes.
    Admin users can see all devices, while regular users only see their own.
    """
    try:
        # Verify tag exists
        tag = await factory.tag_service.get_tag_by_id(tag_id)
        if not tag:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND, 
                detail=f"Tag with ID {tag_id} not found"
            )
        
        # Only return devices owned by the current user unless they're an admin
        owner_filter = None if principal.is_admin else principal.id
        
        devices = await factory.tag_service.get_devices_for_tag(
            tag_id=tag_id, 
            owner_id=owner_filter,
            skip=skip, 
            limit=limit
        )
        
        action = "list_all" if principal.is_admin else "list_own"
        logger.info(f"Access granted: user={principal.username}, resource=tag:{tag_id}/devices, action={action}, count={len(devices)}")
        return devices
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting devices for tag {tag_id}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error getting devices for tag: {str(e)}"
        )

@router.post("/assign", status_code=status.HTTP_200_OK)
async def assign_tags_to_devices(
    assignment: TagAssignment,
    principal: UserPrincipal = Depends(require_scope(["write:devices", "write:tags"])),
    session: AsyncSession = Depends(get_async_session),
    factory: ServiceFactory = Depends(ServiceFactory)
) -> Dict[str, Any]:
    """
    Assign multiple tags to multiple devices in bulk.
    
    This endpoint requires write:devices and write:tags scopes.
    User must be admin or own all devices.
    """
    try:
        # Check device ownership or admin status
        if not principal.is_admin:
            # Verify user owns all devices
            for device_id in assignment.device_ids:
                device = await factory.device_service.get_device_by_id(device_id)
                if not device:
                    raise HTTPException(
                        status_code=status.HTTP_404_NOT_FOUND,
                        detail=f"Device with ID {device_id} not found"
                    )
                if device.owner_id != principal.id:
                    raise HTTPException(
                        status_code=status.HTTP_403_FORBIDDEN,
                        detail="You don't have permission to modify one or more devices"
                    )
        
        # Verify all tags exist
        for tag_id in assignment.tag_ids:
            tag = await factory.tag_service.get_tag_by_id(tag_id)
            if not tag:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail=f"Tag with ID {tag_id} not found"
                )
        
        # Perform bulk assignment
        result = await factory.tag_service.assign_tags_to_devices(
            assignment.tag_ids, 
            assignment.device_ids
        )
        
        logger.info(f"Tags assigned to devices: tags={len(assignment.tag_ids)}, devices={len(assignment.device_ids)}, user={principal.username}")
        return {
            "success": True,
            "assigned_count": result.get("assigned_count", 0),
            "message": f"Successfully assigned {len(assignment.tag_ids)} tags to {len(assignment.device_ids)} devices"
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error assigning tags to devices: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error assigning tags to devices: {str(e)}"
        )

@router.post("/unassign", status_code=status.HTTP_200_OK)
async def unassign_tags_from_devices(
    removal: TagRemoval,
    principal: UserPrincipal = Depends(require_scope(["write:devices", "write:tags"])),
    session: AsyncSession = Depends(get_async_session),
    factory: ServiceFactory = Depends(ServiceFactory)
) -> Dict[str, Any]:
    """
    Remove multiple tags from multiple devices in bulk.
    
    This endpoint requires write:devices and write:tags scopes.
    User must be admin or own all devices.
    """
    try:
        # Check device ownership or admin status
        if not principal.is_admin:
            # Verify user owns all devices
            for device_id in removal.device_ids:
                device = await factory.device_service.get_device_by_id(device_id)
                if not device:
                    raise HTTPException(
                        status_code=status.HTTP_404_NOT_FOUND,
                        detail=f"Device with ID {device_id} not found"
                    )
                if device.owner_id != principal.id:
                    raise HTTPException(
                        status_code=status.HTTP_403_FORBIDDEN,
                        detail="You don't have permission to modify one or more devices"
                    )
        
        # Perform bulk removal
        result = await factory.tag_service.remove_tags_from_devices(
            removal.tag_ids, 
            removal.device_ids
        )
        
        logger.info(f"Tags removed from devices: tags={len(removal.tag_ids)}, devices={len(removal.device_ids)}, user={principal.username}")
        return {
            "success": True,
            "removed_count": result.get("removed_count", 0),
            "message": f"Successfully removed {len(removal.tag_ids)} tags from {len(removal.device_ids)} devices"
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error removing tags from devices: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error removing tags from devices: {str(e)}"
        )